src/charm.py

```diff
@@ -236,16 +240,93 @@ def _command(self) -> str:
+                self.unit.status = BlockedStatus(f"Error calculating retention size: {e}")
+            except LightkubeApiError as e:
+                self.unit.status = BlockedStatus(
+                    f"Error calculating retention size "
+                    f"(try running `juju trust` on this application): {e}"
+                )
+            else:
+                logger.debug("Retention size limit set to %s (%s%%)", capacity, ratio * 100)
+                args.append(f"--storage.tsdb.retention.size={capacity}")
 
         command = ["/bin/prometheus"] + args
 
         return " ".join(command)
 
-    def _is_valid_timespec(self, timeval) -> bool:
+    def _convert_k8s_capacity_to_legacy_binary_gigabytes(self, capacity: str, multiplier) -> str:
+        # Reduce possible ambiguity in unit name by adding a trailing 'B'.
+        # (bitmath doesn't accept e.g. "Gi" because it also supports bits.)
+        if not capacity.endswith("B"):
+            capacity += "B"
+
+        # For simplicity, always communicate to prometheus in GiB
+        storage_value = multiplier * bitmath.parse_string(capacity).to_GiB().value
+        return f"{storage_value}GB"
+
+    def _get_pvc_capacity(self) -> str:
+        """Get PVC capacity from pod name.
+
+        This may need to be handled differently once Juju supports multiple storage instances
+        for k8s (https://bugs.launchpad.net/juju/+bug/1977775).
+        """
+        # This assertion would be picked up by every integration test so no concern this would
+        # reach production.
+        assert (
+            "database" in self.model.storages
+        ), "The 'database' storage is no longer in metadata: must update literals in charm code."
+
+        # Get PVC capacity from kubernetes
+        client = Client()
+        pod_name = self.unit.name.replace("/", "-", -1)
+
+        # Take the first volume whose name starts with "<app-name>-database-".
+        # The volumes array looks as follows for app "am" and storage "data":
+        # 'volumes': [{'name': 'am-data-d7f6a623',
+        #              'persistentVolumeClaim': {'claimName': 'am-data-d7f6a623-am-0'}}, ...]
+        pvc_name = ""
+        for volume in cast(
+            Pod, client.get(Pod, name=pod_name, namespace=self.model.name)
+        ).spec.volumes:
+            # claimName looks like this: 'prom-database-325a0ee8-prom-0'
+            matcher = re.compile(f"^{self.app.name}-database-.*?-{pod_name}$")
+            if matcher.match(volume.persistentVolumeClaim.claimName):
+                pvc_name = volume.persistentVolumeClaim.claimName
+                break
+
+        if not pvc_name:
+            raise ValueError("No PVC found for pod " + pod_name)
+
+        capacity = cast(
+            PersistentVolumeClaim,
+            client.get(PersistentVolumeClaim, name=pvc_name, namespace=self.model.name),
+        ).status.capacity[
+            "storage"  # Storage name must match metadata.yaml
```

Comment:
This comment is incorrectly placed. Fixed in next commit.

Cannot look it up dynamically, unless assuming there is only one:
```yaml
storage:
  database:
    type: filesystem
```

Reply:
If it's actually the parent (`storage`), then the key is fixed anyway.

Otherwise, if it "must match" and it's an actual key such as `database:`, it's in `charm.meta.storages`

`storage` seems concrete enough. At first glance, it looked like a placeholder